### PR TITLE
Define a custom kind type for AConstr

### PR DESCRIPTION
**Pull request by SkySkimmer**
The aconstr kind is used to describe the subterms to recurse on. It has no need to distinguish between term kinds.

This lets us avoid building dummy Evar kinds. We also avoid building aconstr and instead get_max_rel for subterms that don't get recursed on.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@coqbot bench

**Review by ppedrot — APPROVED** [review]
Clever trick (if it works).

File: pretyping/unification.ml [hunk 1]
```diff
@@ -1960,64 +1961,17 @@ let w_unify_to_subterm env evd ?(flags=default_unify_flags ()) (op,cl) =
...
-              iter_fail matchrec t)
-
-          | Cast (_, _, _) (* Is this expected? *)
-          | Rel _ | Var _ | Meta _ | Evar _ | Sort _ | Const _ | Ind _
-            | Construct _ | Int _ | Float _ -> user_err Pp.(str "Match_subterm")))
+        | ACast _ -> assert false (* just got stripped *)
+        | AApp (f,args) ->
+          let n = Array.length args in
+          assert (n>0);
+          let c1 = AConstr.mkApp (f,Array.sub args 0 (n-1)) in
```
Comment (ppedrot) [inline pretyping/unification.ml:1968]: With this simpler representation, we could even try to be a bit more clever and not reconstruct the subarrays here.

**Comment by ppedrot** [comment]
No clear gain from the bench but I find the code much cleaner, so let's merge. @coqbot merge now

**Pull request merged**